src/persist/src/indexed/mod.rs

```diff
@@ -364,48 +373,25 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
         Ok(())
     }
 
-    /// Returns the current "sealed" frontier for an id.
-    ///
-    /// This frontier represents a contract of time such that all updates with a
-    /// time less than it have arrived. This frontier is advanced though the
-    /// `seal` method. Once a time has been sealed for an id, it becomes an
-    /// error to later seal it at an time less than or equal to the sealed
-    /// frontier. It is also an error to write new data with a time less than
-    /// the sealed frontier.
-    pub fn sealed_frontier(&mut self, id: Id) -> Result<Antichain<u64>, Error> {
-        let trace = self
-            .traces
-            .get_mut(&id)
-            .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
-        Ok(trace.ts_upper())
-    }
-
-    /// Atomically moves all writes in future not in advance of the given
-    /// timestamp into the trace and does any necessary resulting compaction
-    /// work.
-    ///
-    /// Sealing a time advances the "sealed" frontier for an id, which restricts
-    /// what times can later be sealed and written for that id. See
-    /// `sealed_frontier` for details.
-    pub fn seal(&mut self, ids: Vec<Id>, ts_upper: u64) -> Result<(), Error> {
-        // TODO: Separate the logical work of seal which just disallows future
-        // updates and seals at times <= ts_upper from the physical work of
-        // moving things from the future to the trace. This could let us
-        // amortize the work of doing so across frequent seal calls? All the
-        // physical movement could live in `step`.
+    /// Atomically moves all writes in future not in advance of the trace's
+    /// seal frontier into the trace and does any necessary resulting eviction
+    /// work to remove uneccessary batches.
+    fn drain_future(&mut self) -> Result<(), Error> {
+        for (id, trace) in self.traces.iter_mut() {
+            // If this future is already properly sealed then we don't need
+            // to do anything.
+            let seal = trace.seal();
+            let trace_upper = trace.ts_upper();
+            if seal == trace_upper {
+                continue;
+            }
 
-        for id in ids.into_iter() {
             let future = self
                 .futures
                 .get_mut(&id)
                 .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
-            let trace = self
-                .traces
-                .get_mut(&id)
-                .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
 
-            let batch_upper = Antichain::from_elem(ts_upper);
-            let desc = Description::new(trace.ts_upper(), batch_upper, trace.since());
+            let desc = Description::new(trace_upper, seal, trace.since());
```

Comment:
Not this change, but I think we actually want an antichain of the minimal element for the since here. since we don't compact in future, a new batch always has the full fidelity of the original data

Reply:
done